### PR TITLE
Add UpdateProductShipping command

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Implement command to update product shipping options. Cover basic use-cases with behat scenarios
| Type?         | refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | yes
| Fixed ticket? | part of #14774
| How to test?  | none

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/20107)

### CODE REVIEW

File: tests/Integration/Behaviour/Features/Context/Domain/CarrierFeatureContext.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+use Carrier;
+use Configuration;
+use Context;
+use Group;
+use PrestaShopException;
+
+class CarrierFeatureContext extends AbstractDomainFeatureContext
+{
+    /**
+     * @todo: It is a temporary method to use sharedStorage and should be improved once Carrier creation is migrated.
```
Comment (matks) [inline tests/Integration/Behaviour/Features/Context/Domain/CarrierFeatureContext.php:40]: 👍

File: tests/Integration/Behaviour/Features/Context/Domain/ProductFeatureContext.php [hunk 2]
```diff
@@ -467,6 +559,12 @@ public function assertProductFields(string $productReference, TableNode $table)
 
         $this->assertTaxRulesGroup($data, $productForEditing);
         $this->assertPriceFields($data, $productForEditing->getPricesInformation());
+        $this->assertShippingInformation($data, $productForEditing->getShippingInformation());
+
+        // Assertions checking isset() can hide some errors if it doesn't find array key,
```
Comment (matks) [inline tests/Integration/Behaviour/Features/Context/Domain/ProductFeatureContext.php:569]: 👍

**Review by matks — DISMISSED** [review]
Code is good. But I discover the concept of "delivery time note". What is this thing ? 😅

**Comment by zuk3975** [comment]
> Code is good. But I discover the concept of "delivery time note". What is this thing ?

Basically i just used another name for legacy properties to reflect the use-case.

There are following properties in Product obj model (and database):
1. $additional_delivery_times
2. $delivery_in_stock
3. $delivery_out_stock

By the names you couldn't guess what are their purpose.
As I found out the `2` and `3` are some kind of notes, which appears in front office product details and should tell delivery time depending if product is in stock or out of stock (eg: Delivered within 3-4 days).
The `1` one turns out to be a setting, which tells if those above-mentioned notes should be specific for this product (which are the `2` and `3`, or taken from default settings (`admin-dev/configure/shop/product-preferences/` -> `Product stock`), or turned off.

So in domain i used abit tweaked names to better reflect their use-case
I called them accordingly:
1. $deliveryTimeNotesType
2. $deliveryTimeInStockNotes
3. $deliveryTimeOutOfStockNotes

EDIT: @matks, wdyt ? (You offered some similar renaming regarding "notes" in slack chat, do you still think I could rename it to something more relevant?)

File: src/Adapter/Product/CommandHandler/UpdateProductShippingHandler.php [hunk 3]
```diff
@@ -0,0 +1,119 @@
+<?php
+/**
+ * 2007-2020 PrestaShop SA and Contributors
```
Comment (Progi1984) [inline src/Adapter/Product/CommandHandler/UpdateProductShippingHandler.php:3]: Please use new license headers

**Review by Progi1984 — COMMENTED** [review]
Check comments

File: src/Adapter/Product/CommandHandler/UpdateProductShippingHandler.php (hunk 3 above)
Comment (zuk3975) [reply src/Adapter/Product/CommandHandler/UpdateProductShippingHandler.php:3]: I've updated headers in my editor. 
Not gonna update them manually now, as Im aware there is a command for that, which will be run at some point later to update all headers isn't it?

Comment (matks) [reply src/Adapter/Product/CommandHandler/UpdateProductShippingHandler.php:3]: Indeed. @Progi1984 I regularly run the command to update the codebase and make sure it stays fine 😉

**Review by matks — APPROVED** [review]

File: tests/Integration/Behaviour/Features/Context/Domain/ProductFeatureContext.php [hunk 4]
```diff
@@ -257,6 +261,94 @@ public function updateProductOptions(string $productReference, TableNode $table)
...
+     *
+     * @return array values that was provided, but wasn't handled
+     */
+    private function setUpdateShippingCommandData(array $data, UpdateProductShippingCommand $command): array
+    {
+        $unhandledValues = $data;
+
+        if (isset($data['width'])) {
+            $command->setWidth($data['width']);
+            unset($unhandledValues['width']);
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/ProductFeatureContext.php:302]: Oohh, that's a really nice improvement We should add this kind of checks in other tests 👍

File: tests/Integration/Behaviour/Features/Context/Domain/ProductFeatureContext.php [hunk 5]
```diff
@@ -472,6 +564,12 @@ public function assertProductFields(string $productReference, TableNode $table)
 
         $this->assertTaxRulesGroup($data, $productForEditing);
         $this->assertPriceFields($data, $productForEditing->getPricesInformation());
+        $this->assertShippingInformation($data, $productForEditing->getShippingInformation());
+
+        // Assertions checking isset() can hide some errors if it doesn't find array key,
+        // to make sure all provided fields were checked we need to unset every asserted field
+        // and finally, if provided data is not empty, it means there are some unnasserted values left
+        Assert::assertEmpty($data, sprintf('Some provided fields haven\'t been asserted: %s', implode(',', $data)));
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/ProductFeatureContext.php:572]: Nice one
I'm only afraid of one thing, in case you need to perform two checks on one field (or use it for anything else) it will have been cleared from the `$data` This might lead to unexpected behaviour for someone who is not aware of this cleaning (like "why did my data disappear?")

But for now, as long as we don't have issues this is perfect for the job

File: tests/Integration/Behaviour/Features/Scenario/CLDR/cldr.feature [hunk 6]
```diff
@@ -1,5 +1,6 @@
 # ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml -s cldr
 @reset-database-before-feature
+@clear-cache-after-feature
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/CLDR/cldr.feature:3]: Why did you need to change this? Did it impact the product tests?
I'm always very careful with tests related to CLDR.. I don't know why 😅

File: tests/Integration/Behaviour/Features/Scenario/Product/update_shipping.feature [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+      | height                           | 5                          |
+      | depth                            | 4                          |
+      | weight                           | 2                          |
+
+  Scenario: Provide negative additional shipping cost
+    Given product product1 should have following values:
+      | additional_shipping_cost         | 12                         |
+    When I update product product1 shipping information with following values:
+      | additional_shipping_cost         | -12                        |
+    Then I should get error that product additional_shipping_cost is invalid
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Product/update_shipping.feature:100]: You can add a check that `additional_shipping_cost` is still 12 like you did for other fields

**Review by jolelievre — APPROVED** [review]

**Comment by jolelievre** [comment]
Thank you @zuk3975

File: tests/Integration/Behaviour/Features/Scenario/CLDR/cldr.feature (hunk 6 above)
Comment (zuk3975) [reply tests/Integration/Behaviour/Features/Scenario/CLDR/cldr.feature:3]: for a record, 
It used to cache languages in memory, so my product multilang scenarios failed. 
e.g. I was expecting only EN and FR langs, but there was multiple other languages created, so values didn't match.

File: tests/Integration/Behaviour/Features/Context/Domain/ProductFeatureContext.php (hunk 5 above)
Comment (zuk3975) [reply tests/Integration/Behaviour/Features/Context/Domain/ProductFeatureContext.php:572]: Yea, we'll figure something out when time comes :sweat_smile:
